WIP: Position control landing and takeoff

MaEtUgR opened the pull request:
This pull request suggests subtle but fundamental changes to the position controller and landing detector to achieve an intuitive all the time manual controlled takeoff and landing of a multicopter.

Summary:
- hover throttle is used as feed forward term against constant gravitational force
  (misusing the thrust integration control for this purpose caused problems)
- land detector keeps vehicle landed until a mode dependent manual throttle input is exceeded
  (this unlocks position control takeoff with spring loaded throttle sticks)
- good refactor commit by @julianoes 
- thrust amount calculation in position control takes 3D projection onto the real vehicle thrust direction
  (useful in general but added here especially because horizontal position control corrections while landing produced vertical thrust and didn't allow proper landing)

Foremost the last point still needs real world testing.

Stifael commented on src/modules/land_detector/MulticopterLandDetector.cpp:
```diff
@@ -230,5 +237,18 @@ bool MulticopterLandDetector::_get_landed_state()
+	}
+
+	return 0.0f;
+}
```
can you specify what 0.9 and 0.15 is and where the number come from. I realize that 0.15 was already used before, but since this value somewhat represent min and max, we could also label them instead of magic numbers

MaEtUgR replied:
0.15 was there, it's minimal throttle to start a takeoff in manual flight mode.
I chose 0.9 to be useful when taking off in position mode. You don't accidentally hit this throttle point and it gives you a good takeoff velocity such that you come quickly of the ground. You are right, defined constants would be better. I'll fix that.

Stifael commented on src/modules/mc_pos_control/mc_pos_control_main.cpp:
```diff
@@ -568,6 +568,7 @@ MulticopterPositionControl::parameters_update(bool force)
 		param_get(_params_handles.thr_max, &_params.thr_max);
 		param_get(_params_handles.thr_hover, &_params.thr_hover);
+		_params.thr_hover = math::constrain(_params.thr_hover, _params.thr_min, _params.thr_max);
 		param_get(_params_handles.alt_ctl_dz, &_params.alt_ctl_dz);
```
I know the purpose of this pull request is not to change any convention. But it would be great to keep in mind that we are planing on using the matrix lib in the near future.

Stifael replied:
is it a constrain? I think first it needs to be checked that params.thr_min is indeed smaller than param.thr_max. And if para.thr_hover is larger than param.thr_max and param.thr_min, do we really want to set it to max? I would say the value param.thr_hover is then just set incorrectly and I would set it to the mean of max and min

MaEtUgR replied:
I totally support migration to the matrix library. That's why I converted the dot product calculation to the matrix library convention such that it will be compatible to the migration. Nevertheless I think min, max, constrain ... is not purpose of the matrix library.

MaEtUgR replied:
I just wanted to make sure the user is not able to set 10'000% hover throttle because the integrator would not be able to compensate for that and the vehicle would fly to "infinite" altitude. Normal values like 100% will not lead to nice takeoffs but should not cause any flyaway or crash. This matches the level of checks that the rest of the code has.

Stifael commented on src/modules/mc_pos_control/mc_pos_control_main.cpp:
```diff
@@ -1778,7 +1758,8 @@ MulticopterPositionControl::control_position(float dt)
-			thrust_sp = vel_err.emult(_params.vel_p) + _vel_err_d.emult(_params.vel_d) + _thrust_int;
+			thrust_sp = vel_err.emult(_params.vel_p) + _vel_err_d.emult(_params.vel_d)
+				    + _thrust_int - math::Vector<3>(0, 0, _params.thr_hover);
 		}
```
please use the explicit float version: 0.0f instead of the integer 0

MaEtUgR replied:
Makes no difference for the compiled output but for correctness you are totally right.

Stifael commented on src/modules/mc_pos_control/mc_pos_control_main.cpp:
```diff
@@ -1925,10 +1906,15 @@ MulticopterPositionControl::control_position(float dt)
+		matrix::Vector3f R_z(_R(0, 2), _R(1, 2), _R(2, 2));
+		matrix::Vector3f F(thrust_sp.data);
+		thrust_body_z = F.dot(-R_z); /* recalculate because it might have changed */
 
```
why is it a projection onto the negative body-z axis?

MaEtUgR replied:
A normal multicopter can only generate total thrust in body z direction.
I added some comments explaining the calculation I did.

Stifael replied:
Well, or in the negative body z direction as in our case

LorenzMeier commented:
@PX4TestFlights This should include the standard mission test but most importantly takeoff and landing in position control mode.

julianoes commented on src/modules/land_detector/MulticopterLandDetector.cpp:
```diff
@@ -230,5 +237,25 @@ bool MulticopterLandDetector::_get_landed_state()
+		/* Should be above 0.5 because below that we do not gain altitude and won't take off.
+		 * Also it should be quite high such that we don't accidentally take off when using
+		 * a spring loaded throttle and have a useful vertical speed to start with. */
+		return 0.9f;
```
I vote for `0.75`.

MaEtUgR replied:
Yes, we also found that roughly 0.75 would be better.

julianoes commented on src/modules/land_detector/MulticopterLandDetector.cpp:
```diff
@@ -230,5 +237,25 @@ bool MulticopterLandDetector::_get_landed_state()
+	/* Manual/attitude mode */
+	if (_control_mode.flag_control_manual_enabled && _control_mode.flag_control_attitude_enabled) {
+		/* Should be quite low and certainly below hover throttle because pilot controls throttle manually. */
+		return 0.15f;
```
And `0.25` here.

MaEtUgR replied:
Why exactly 0.25?
I took 0.15 because of this check that existed before:
https://github.com/PX4/Firmware/pull/6205/files#diff-bcb38f37ec5bebecc13965e3fec8c312R183

MaEtUgR replied:
Cleared. It's not landing threshold but manual mode takeoff threshold.

julianoes commented on src/modules/mc_pos_control/mc_pos_control_main.cpp:
```diff
@@ -1952,15 +1939,17 @@ MulticopterPositionControl::control_position(float dt)
 				/* Z component is negative, going down, simply limit thrust vector */
-				float k = thr_max / thrust_abs;
+				float k = thr_max / thrust_body_z;
 				thrust_sp *= k;
```
Is `thrust_sp` still used somewhere? I'm confused here.

MaEtUgR replied:
That's an error, thanks for spotting.
I just saw that there's something fishy with https://github.com/PX4/Firmware/pull/6205/files#diff-fc77c3ef569029d45764664c75d4b0c1R1793.

MaEtUgR replied:
The only error was the line I mentioned above but it had no functional effect.
Yes, `thrust_sp` gets used later on in the attitude setpoint calculation.
I did not rework the entire limiting logic yet but at least it works unaltered like before.

MaEtUgR commented:
- Rebased on master
- Changed the two thumb upped things in @julianoes review with history rewrite, force push
- added a commit with a raised minimum throttle threshold
  because of testing experience where landing was not detected

My findings during real world tests: It generally works and is a good first solution.
It's not perfect yet, foremost the landing detection needs some further analysis. Landing not detected can occur in some rare situations but at least no false positives.

francisco3dr commented:
2 Flights

http://logs.px4.io/plot_app?log=daa57100-6845-4dbb-9c1b-24e5c2e92195
http://logs.px4.io/plot_app?log=7491753f-d123-4fc6-9a1b-46a14ca4a0e7

LorenzMeier commented:
@MaEtUgR Note the altitude controller oscillation when the vehicle was hovering. We might need to scale the current default gain. We should *not* require users to re-tune their system.

MaEtUgR commented:
![Actuator output while landing](https://cloud.githubusercontent.com/assets/4668506/21604727/2c618d9e-d1a5-11e6-81c0-b31512d9ad87.png)
I see the landing not detected problem, I also had that in my tests. The land detection needs a guaranteed throttle ramp down when on the ground. This should be realized in a two stage manner. If certain conditions are met we assume we are landed and ramp down the motor thrusts such that we could still recover. And in a second stage when the thrusts are low and nothing changed we are landed for sure. Working on it now.

About the hover problem: I don't really get where I can see it. The PR should theoretically not change anything in steady state hovering but might potentially interfere there and if it does I have to find the cause.

tumbili commented:
@MaEtUgR I'll give this a test on my beloved Bebop.
I already tested in SITL and just have one comment:
I think hover thrust is applied instantaneously once takeoff is detected which I find too abrupt.

The rest seems to work as expected, will report back with a log.

Stifael commented:
I tested this pr in sitl with uneven ground, and in contrast to the current masters I was not able to reproduce the "not" detected landing. This is mainly due to the correct thrust projection onto the desired thrust axis. In addition, in simulation the hoverthrust is very close to the approximated 0.5 hover thrust and therefore landing works better than in reality where the two thrust values differ more. @MaEtUgR is it not possible to split this pr into two pr's? One for the projection onto the desired thrust vector which does improve landing and flight for quads, and another pr for the  takeoff with the hoverthrust as feedforward? And eventually we will have a third for better landing detection

LorenzMeier commented:
Can you do a few trials with an incorrectly set thrust param and see if that reproduces it?

santiago3dr commented:
noticed this PR was still open, here are a couple flights taking off and landing in Position mode

http://logs.px4.io/plot_app?log=312af812-9dfb-4884-a442-6aa6e007cea7
http://logs.px4.io/plot_app?log=ad824ea3-489e-4bc5-8c66-e9a585179eb8

Stifael commented:
I tried to set different thrust hover values, but nothing made a difference. I looked at the landing detection, and all flags are correctly set when threshold is reached. In addition, the hysteresis is doing the correct thing  where states are only changed after 2seconds. 
I also looked again at http://logs.px4.io/plot_app?log=daa57100-6845-4dbb-9c1b-24e5c2e92195, and I dont think that the actuator outputs / thrust values are the issue, but rather they are increasing due to setpoint changes in roll, pitch and yaw caused by position estimate drift:
![x-y_drift](https://cloud.githubusercontent.com/assets/13449161/21764745/f218ab6c-d664-11e6-89c7-41784727911a.png). I also dont get wht gps altitude inreases as well.

LorenzMeier commented:
Done in https://github.com/PX4/Firmware/pull/6327.
